File: tests/pluginTests.ts
import { expect } from 'chai';
import PluginManager from '../src/core/Infrastructure';
import MyPlugin from '../src/plugins/MyPlugin';

describe('Plugin Architecture Tests', () => {
    let pluginManager: PluginManager;

    beforeEach(() => {
        pluginManager = new PluginManager();
        pluginManager.register(new MyPlugin());
    });

    it('should register a plugin successfully', () => {
        expect(pluginManager).to.have.property('plugins').with.lengthOf(1);
    });

    it('should execute the plugin and return expected result', async () => {
        const results = await pluginManager.executeAll({ key: 'value' });
        expect(results[0]).to.deep.equal({ success: true, data: { key: 'value' } });
    });

    it('should destroy all plugins successfully', () => {
        pluginManager.destroyAll();
        expect(pluginManager).to.have.property('plugins').with.lengthOf(0);
    });
});
import { PluginInterface } from '../src/plugins/PluginInterface';
import PluginManager from '../src/core/Infrastructure';

class MockPlugin implements PluginInterface {
    name = 'MockPlugin';
    version = '1.0.0';
    description = 'Bir test eklenti';
    permissions = ['execute'];

    init(): void {
        console.log(`${this.name} başlatıldı.`);
    }

    async execute(params: any): Promise<any> {
        return { success: true, data: params };
    }

    destroy(): void {
        console.log(`${this.name} sonlandırıldı.`);
    }
}

describe('PluginManager', () => {
    let pluginManager: PluginManager;

    beforeEach(() => {
        pluginManager = new PluginManager();
    });

    it('should register a valid plugin', () => {
        const mockPlugin = new MockPlugin();
        expect(() => pluginManager.register(mockPlugin)).not.toThrow();
    });

    it('should execute all registered plugins', async () => {
        const mockPlugin1 = new MockPlugin();
        const mockPlugin2 = new MockPlugin();

        pluginManager.register(mockPlugin1);
        pluginManager.register(mockPlugin2);

        const results = await pluginManager.executeAll({ key: 'value' });
        expect(results).toHaveLength(2);
        expect(results[0]).toEqual({ success: true, data: { key: 'value' } });
        expect(results[1]).toEqual({ success: true, data: { key: 'value' } });
    });

    it('should handle plugin execution errors', async () => {
        class FaultyPlugin implements PluginInterface {
            name = 'FaultyPlugin';
            version = '1.0.0';
            description = 'Hata veren bir eklenti';
            permissions = ['execute'];

            init(): void {}
            async execute(params: any): Promise<any> {
                throw new Error('Execution failed');
            }
            destroy(): void {}
        }

        const faultyPlugin = new FaultyPlugin();
        pluginManager.register(faultyPlugin);

        const results = await pluginManager.executeAll({ key: 'value' });
        expect(results).toHaveLength(1);
        expect(results[0]).toHaveProperty('success', false);
        expect(results[0]).toHaveProperty('error');
    });

    it('should destroy all registered plugins', () => {
        const mockPlugin1 = new MockPlugin();
        const mockPlugin2 = new MockPlugin();

        pluginManager.register(mockPlugin1);
        pluginManager.register(mockPlugin2);

        console.log = jest.fn(); // Suppress console output for this test
        pluginManager.destroyAll();
        expect(console.log).toHaveBeenCalledTimes(2);
    });
});
import { PluginInterface } from '../plugins/PluginInterface';
import PluginManager from '../core/Infrastructure';

class MockPlugin implements PluginInterface {
    name = 'MockPlugin';
    version = '1.0.0';
    description = 'Bir test eklenti';
    permissions = ['execute'];

    init(): void {
        console.log(`${this.name} başlatıldı.`);
    }

    async execute(params: any): Promise<any> {
        return { success: true, data: params };
    }

    destroy(): void {
        console.log(`${this.name} sonlandırıldı.`);
    }
}

class FaultyPlugin implements PluginInterface {
    name = 'FaultyPlugin';
    version = '1.0.0';
    description = 'Hata veren bir eklenti';
    permissions = ['execute'];

    init(): void {}
    async execute(params: any): Promise<any> {
        throw new Error('Execution failed');
    }
    destroy(): void {}
}

describe('PluginManager', () => {
    let pluginManager: PluginManager;

    beforeEach(() => {
        pluginManager = new PluginManager();
    });

    it('should register a valid plugin', () => {
        const mockPlugin = new MockPlugin();
        expect(() => pluginManager.register(mockPlugin)).not.toThrow();
    });

    it('should handle plugin execution errors', async () => {
        const faultyPlugin = new FaultyPlugin();
        pluginManager.register(faultyPlugin);
        const results = await pluginManager.executeAll({ key: 'value' });
        expect(results[0].success).toBe(false);
        expect(results[0].error.message).toBe('Execution failed');
    });

    it('should not execute a plugin without required permissions', async () => {
        class NoPermissionPlugin implements PluginInterface {
            name = 'NoPermissionPlugin';
            version = '1.0.0';
            description = 'İzinleri olmayan bir eklenti';
            permissions: string[] = [];

            init(): void {}
            async execute(params: any): Promise<any> {
                return { success: true, data: params };
            }
            destroy(): void {}
        }

        const noPermissionPlugin = new NoPermissionPlugin();
        pluginManager.register(noPermissionPlugin);
        await expect(pluginManager.executeAll({ key: 'value' }, ['execute'])).rejects.toThrow('Plugin NoPermissionPlugin does not have the required permissions.');
    });
});
````

File: tests/pluginTests.py
import os
import unittest
from src.core.Infrastructure import PluginManager
from src.plugins.PluginInterface import PluginInterface

class MockPlugin(PluginInterface):
    name = 'MockPlugin'
    version = '1.0.0'
    description = 'Bir test eklenti'
    permissions = ['execute']

    def init(self) -> None:
        print(f"{self.name} başlatıldı.")

    async def execute(self, params: any) -> any:
        if hasattr(self, 'sanitizeInputs'):
            self.sanitizeInputs(params)
        return { "success": True, "data": params }

    def destroy(self) -> None:
        print(f"{self.name} sonlandırıldı.")

    def sanitizeInputs(self, params: any) -> None:
        # Girdi verilerini temizleme mantığınızı buraya yazın
        if not isinstance(params, dict):
            raise ValueError('Geçersiz girdi parametreleri.')
        for key in params:
            value = params[key]
            if isinstance(value, str) and ("'" in value or ";" in value):
                raise ValueError('Potansiyel güvenlik ihlali tespit edildi.')

class FaultyPlugin(PluginInterface):
    name = 'FaultyPlugin'
    version = '1.0.0'
    description = 'Hata veren bir eklenti'
    permissions = ['execute']

    def init(self) -> None:
        pass

    async def execute(self, params: any) -> any:
        raise Exception("Execution failed")

    def destroy(self) -> None:
        pass

class TestPluginManager(unittest.TestCase):
    def setUp(self) -> None:
        self.plugin_manager = PluginManager()

    def test_register_plugin(self) -> None:
        mock_plugin = MockPlugin()
        self.plugin_manager.register(mock_plugin)
        self.assertIn(mock_plugin, self.plugin_manager.plugins)

    def test_execute_plugin_success(self) -> None:
        mock_plugin = MockPlugin()
        self.plugin_manager.register(mock_plugin)
        result = self.plugin_manager.executeAll({"key": "value"})
        self.assertTrue(result[0]["success"])
        self.assertEqual(result[0]["data"], {"key": "value"})

    def test_execute_plugin_failure(self) -> None:
        faulty_plugin = FaultyPlugin()
        self.plugin_manager.register(faulty_plugin)
        result = self.plugin_manager.executeAll({"key": "value"})
        self.assertFalse(result[0]["success"])
        self.assertIsNotNone(result[0]["error"])

    def test_sandbox_environment(self) -> None:
        mock_plugin = MockPlugin()
        self.plugin_manager.register(mock_plugin)
        try:
            with unittest.mock.patch.dict(os.environ, {"SANDBOX_ENABLED": "false"}):
                result = self.plugin_manager.executeAll({"key": "value"})
                self.assertFalse(result[0]["success"])
                self.assertIsNotNone(result[0]["error"])
        except Exception as e:
            self.fail(f"Test başarısız oldu: {e}")

    def test_plugin_permissions(self) -> None:
        mock_plugin = MockPlugin()
        faulty_plugin = FaultyPlugin()

        # Test with insufficient permissions
        try:
            result = self.plugin_manager.executeAll({"key": "value"}, ['read'])
            self.assertFalse(result[0]["success"])
            self.assertIsNotNone(result[0]["error"])
        except Exception as e:
            self.fail(f"Test başarısız oldu: {e}")

        # Test with sufficient permissions
        try:
            result = self.plugin_manager.executeAll({"key": "value"}, ['execute'])
            self.assertTrue(result[0]["success"])
            self.assertEqual(result[0]["data"], {"key": "value"})
        except Exception as e:
            self.fail(f"Test başarısız oldu: {e}")

    def test_plugin_sanitize_inputs(self) -> None:
        mock_plugin = MockPlugin()
        self.plugin_manager.register(mock_plugin)
        try:
            with unittest.mock.patch.object(mock_plugin, 'sanitizeInputs') as mock_method:
                result = self.plugin_manager.executeAll({"key": "value"})
                mock_method.assert_called_once_with({"key": "value"})
                self.assertTrue(result[0]["success"])
                self.assertEqual(result[0]["data"], {"key": "value"})
        except Exception as e:
            self.fail(f"Test başarısız oldu: {e}")

    def test_plugin_sanitize_inputs_invalid(self) -> None:
        mock_plugin = MockPlugin()
        self.plugin_manager.register(mock_plugin)
        try:
            with self.assertRaises(Exception):
                result = self.plugin_manager.executeAll({"key": "value'; DROP TABLE users;"})
        except Exception as e:
            self.fail(f"Test başarısız oldu: {e}")

    def test_plugin_enforce_sandbox(self) -> None:
        mock_plugin = MockPlugin()
        self.plugin_manager.register(mock_plugin)
        try:
            with unittest.mock.patch.dict(os.environ, {"SANDBOX_ENABLED": "true"}):
                result = self.plugin_manager.executeAll({"key": "value"})
                self.assertTrue(result[0]["success"])
                self.assertEqual(result[0]["data"], {"key": "value"})
        except Exception as e:
            self.fail(f"Test başarısız oldu: {e}")

    def test_plugin_sanitize_inputs_invalid_command(self) -> None:
        mock_plugin = MockPlugin()
        self.plugin_manager.register(mock_plugin)
        try:
            with self.assertRaises(Exception):
                result = self.plugin_manager.executeAll({"key": "rm -rf /"})
        except Exception as e:
            self.fail(f"Test başarısız oldu: {e}")

    def test_plugin_sanitize_inputs_invalid_sql(self) -> None:
        mock_plugin = MockPlugin()
        self.plugin_manager.register(mock_plugin)
        try:
            with self.assertRaises(Exception):
                result = self.plugin_manager.executeAll({"key": "SELECT * FROM users;"})
        except Exception as e:
            self.fail(f"Test başarısız oldu: {e}")

    def test_plugin_sanitize_inputs_invalid_script(self) -> None:
        mock_plugin = MockPlugin()
        self.plugin_manager.register(mock_plugin)
        try:
            with self.assertRaises(Exception):
                result = self.plugin_manager.executeAll({"key": "<script>alert('XSS');</script>"})
        except Exception as e:
            self.fail(f"Test başarısız oldu: {e}")

if __name__ == '__main__':
    unittest.main()
